src/app/games/diaballik/DiaballikMoveGenerator.ts

```diff
@@ -5,61 +5,146 @@ import { DiaballikNode, DiaballikRules } from './DiaballikRules';
 import { Player, PlayerOrNone } from 'src/app/jscaip/Player';
 import { Coord } from 'src/app/jscaip/Coord';
 import { Direction, Orthogonal } from 'src/app/jscaip/Direction';
-import { MGPFallible } from 'src/app/utils/MGPFallible';
 import { MGPOptional } from 'src/app/utils/MGPOptional';
+import { Utils } from 'src/app/utils/utils';
+import { MGPFallible } from 'src/app/utils/MGPFallible';
+
+class DiaballikMoveInConstruction {
+
+    public readonly hasPass: boolean;
+    public readonly translations: number;
+
+    public constructor(public readonly subMoves: DiaballikSubMove[],
+                       public readonly stateAfterSubMoves: DiaballikState) {
+        Utils.assert(this.subMoves.length <= 3, 'DiaballikMoveInConstruction can have at most 3 submoves');
+        let hasPass: boolean = false;
+        let translations: number = 0;
+        for (const subMove of subMoves) {
+            if (subMove instanceof DiaballikPass) {
+                Utils.assert(hasPass === false, 'DiaballikMoveInConstruction can have at moste one pass');
+                hasPass = true;
+            } else {
+                Utils.assert(translations < 2, 'DiaballikMoveInConstruction can have at most two translations');
+                translations++;
+            }
+        }
+        this.hasPass = hasPass;
+        this.translations = translations;
+    }
+
+    public addIfLegal(subMove: DiaballikSubMove, listToAddTo: DiaballikMoveInConstruction[]): void {
+        const legality: MGPFallible<DiaballikState> = DiaballikRules.get().isLegalSubMove(this.stateAfterSubMoves, subMove);
+        if (legality.isSuccess()) {
+            const newMoveInConstruction: DiaballikMoveInConstruction =
+                new DiaballikMoveInConstruction(this.subMoves.concat([subMove]), legality.get());
+            listToAddTo.push(newMoveInConstruction);
+        }
+    }
```

Comment:
this n'est pas impliqué, rendre ça statique car là c'est confusant.
P-e renommer addSubMoveToDraftMoveList ?
Draft au lieu de inConstruction en général peut être stylé non ?

Reply:
this est utilisé deux fois (maintenant 3 après refactors), donc j'ai pas changé en statique. Peut-être que renommer peut aider à éviter la confusion ? `addWithNewSubMoveToListIfLegal` ?

hmm draft j'suis pas convaincu, en construction c'est vraiment ce que ça représente : t'es en train de construire le mouvement. T'as pas un mouvement "brouillon", ce qui pourrait impliquer que tu peux changer n'importe quoi dans le draftMove, alors que le moveInConstruction t'y ajoutes juste des trucs.